### PR TITLE
Don't use WINDOWS/POSIX/NOMINMAX macros defined in cmake, move them to source code instead

### PR DESCRIPTION
**Pull request by antkmsft**
We should use `_WIN32` that is defined by platform and compiler, instead of `WINDOWS` that we define in our cmake.

First of all, we absolutely can't do that in public headers: when user code includes that header, things that are defined in our cmake do not apply, so `#ifdef WINDOWS` evaluates as false on Windows, while `#ifdef _WIN32` does what we want correctly.

Second, it requires us to always make sure we always include the DefinePlatform cmake in all our cmake files. It will become more significant and harder to follow once we have multiple "entry points" to build a project, by just building single library directory instead of the whole SDK (vcpkg case).

Third, it is overall goodness to rely less on the fact that our code is being built exclusively in cmake environment. If I recall correctly, it was one of the principles (maybe unwritten) we had in C SDK, meybe in this one too at some point. IF we use `def/ndef _WIN32`, you can build the file in command line, without any project files, and everything would compile as intended when written.

You might ask, why I do `ifndef X define X` - that is to avoid warnings. We can suppress them for .cpp files, but not for the headers (see reason#1 above). But for the uniformness, I use `ifndef X define X` in .cpp files as well. There aren't many places where we have to do this rather verbose trick, so I think it is ok. Plus, less warnings suppressed is better.

--
This change was extracted from #1013, in order to isolate some of the changes.

### CODE REVIEW

File: sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp:11]: here's a thought, macros are not encouraged to be used according to our guidelines. We have `std::min` and `std::max` in `<algorithm>` which is always better than `min` and `max` macros in `Windows.h`.

We have `NOMINMAX` defined in CMakeFile, so this define here is unnecessary

File: sdk/storage/CMakeLists.txt
Comment (Jinming-Hu) [inline sdk/storage/CMakeLists.txt:14]: what's the motivation of removing this?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp:11]: then, okay.

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp
Comment (RickWinter) [inline sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp:92]: We need to check the output of this function call.  If it fails we need to throw,  Failure is denoted by the return code of INVALID_HANDLE_VALUE

File: sdk/core/azure-core/test/ut/transport_adapter_base.cpp
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/transport_adapter_base.cpp:451]: Need to check the open for failure.

**Review by RickWinter — CHANGES_REQUESTED** [review]
You've adopted a pattern of if Windows else everything else.  Instead of that we should be if Windows elsif POSIX else #error .  This way if someone compiles on a platform that does not define windows or posix they will know its not a supported configuration.

File: sdk/storage/CMakeLists.txt
Comment (antkmsft) [reply sdk/storage/CMakeLists.txt:14]: I listed in PR description - it is better not to rely on the fact that our project is always built via cmake, and when we need to enable vcpkg build (#1013), we need individually buildable directories, which means that we can't rely on the fact that, for example, `storage/azure-storage-common/CMakeLists.txt` was loaded from the `storage/CMakeLists.txt` and therefore all the definitions from the broader directory file are present when building azure-storage-common.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp
Comment (antkmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/file_io.hpp:11]: I very much dislike MIN and MAX macros that windows.h defines too, and they are harmful - they break the build, break the valid C++ code, and only present on Windows. But I am not removing them, I am just moving the definition from CMakeFile down to the code, motivation being the one I put in my other comment replying to your question about the motivation other comment.

File: sdk/core/azure-core/test/ut/transport_adapter_base.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/transport_adapter_base.cpp:451]: This is not a goal of this PR, I am just moving things around, let's take care of this in a separate PR. @vhvb1989 will you look into this, should be a simple fix, or should I create a work item?

File: sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp
Comment (antkmsft) [reply sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp:92]: Again, hhis is not a goal of this PR, I am just moving things around, let's take care of this in a separate PR. @vhvb1989 will you look into this, should be a simple fix, or should I create a work item?

**Comment by antkmsft** [comment]
> You've adopted a pattern of if Windows else everything else. Instead of that we should be if Windows elsif POSIX else #error . This way if someone compiles on a platform that does not define windows or posix they will know its not a supported configuration.

Detecting POSIX via macros is not as straightforward as `_WIN32`, and I thought we could get away with `!_WIN32` for years with this, given that there aren't many desktop platforms that are not POSIX and Windows. But I'll see what I can do with this https://stackoverflow.com/questions/11350878/how-can-i-determine-if-the-operating-system-is-posix-in-c. `#error` will be helpful in that case, however if you think something is POSIX when it is not, and you try to compile, you get errors as well, just not as straightforward.

File: sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp:92]: Issue for another PR. 
https://github.com/Azure/azure-sdk-for-cpp/issues/1074

Agree with Anton, this is not related for this fix

**Review by vhvb1989 — APPROVED** [review]
LGTM.
thank you for moving this out of the other PR!

File: sdk/core/azure-core/test/ut/transport_adapter_base.cpp
Comment (RickWinter) [reply sdk/core/azure-core/test/ut/transport_adapter_base.cpp:451]: Although its not a goal of the PR you are touching the line of code that is broken.  Its a best practice that if you touch a line of code you should do your best to ensure that line of code is correct.  I made a bug for it

**Comment by RickWinter** [comment]
> > You've adopted a pattern of if Windows else everything else. Instead of that we should be if Windows elsif POSIX else #error . This way if someone compiles on a platform that does not define windows or posix they will know its not a supported configuration.
> 
> Detecting POSIX via macros is not as straightforward as `_WIN32`, and I thought we could get away with `!_WIN32` for years with this, given that there aren't many desktop platforms that are not POSIX and Windows. But I'll see what I can do with this https://stackoverflow.com/questions/11350878/how-can-i-determine-if-the-operating-system-is-posix-in-c. `#error` will be helpful in that case, however if you think something is POSIX when it is not, and you try to compile, you get errors as well, just not as straightforward.

The goal is code correctly compiles for the target with errors presented at the earliest possible point.  If the target isn't POSIX but we compile as if it is, the resulting binary will have Undefined behavior.  Given we require open to be present in the non-_win32 case, this case might be ok.  However if you can figure out a pattern to identify POSIX, that is ideal

File: sdk/core/azure-core/test/ut/transport_adapter_base.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter_base.cpp:451]: This is tests code.
A file body stream will never open a file, it expects the file to be open already by the time the FileBodyStream is created.

Do we need to worry about this in the tests that are only checking the FileBodyStream behavior and not the file open() functions?

File: sdk/storage/CMakeLists.txt
Comment (Jinming-Hu) [reply sdk/storage/CMakeLists.txt:14]: I think you're right

**Comment by Jinming-Hu** [comment]
> > You've adopted a pattern of if Windows else everything else. Instead of that we should be if Windows elsif POSIX else #error . This way if someone compiles on a platform that does not define windows or posix they will know its not a supported configuration.
> 
> Detecting POSIX via macros is not as straightforward as `_WIN32`, and I thought we could get away with `!_WIN32` for years with this, given that there aren't many desktop platforms that are not POSIX and Windows. But I'll see what I can do with this https://stackoverflow.com/questions/11350878/how-can-i-determine-if-the-operating-system-is-posix-in-c. `#error` will be helpful in that case, however if you think something is POSIX when it is not, and you try to compile, you get errors as well, just not as straightforward.

I agree, I think it's good enough to assume the platform is either Windows or POSIX for now.

**Comment by antkmsft** [comment]
@RickWinter, I am nod detecting POSIX explicitly using this recommendation: https://stackoverflow.com/questions/2989810/which-cross-platform-preprocessor-defines-win32-or-win32-or-win32 (this thread has Billy's reply from 2010 BTW :)) Also I am using `#error` where you suggested.
@Jinming-Hu, I hope you are fine with this new method. It should lead to the same results as in previous iteration.

**Comment by Jinming-Hu** [comment]
> @Jinming-Hu, I hope you are fine with this new method. It should lead to the same results as in previous iteration.

Yeah, it's good. But it's cumbersome to write this long `defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__)` everywhere. Do you consider defining it somewhere?

**Comment by antkmsft** [comment]
> Yeah, it's good. But it's cumbersome to write this long `defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__)` everywhere. Do you consider defining it somewhere?

I understand. Let me think about it. Maybe we can create a header file in core, call it "platform.hpp", and define WINDOWS and POSIX?

**Comment by Jinming-Hu** [comment]
> > Yeah, it's good. But it's cumbersome to write this long `defined(__unix__) || defined(__unix) || (defined(__APPLE__) && defined(__MACH__)` everywhere. Do you consider defining it somewhere?
> 
> I understand. Let me think about it. Maybe we can create a header file in core, call it "platform.hpp", and define WINDOWS and POSIX?

I like this idea. This is even more helpful when we need to differentiate macOS and Linux. Because I know some file I/O APIs have some differences between these two platforms. Further more, we can add more macros here when we decide to support more architectures (at lease storage team has ARM support in backlog).

**Comment by antkmsft** [comment]
> I like this idea. This is even more helpful when we need to differentiate macOS and Linux. Because I know some file I/O APIs have some differences between these two platforms. Further more, we can add more macros here when we decide to support more architectures (at lease storage team has ARM support in backlog).

Now that you described all the possibilities, I like the idea even more. Check out the latest commit. I hope you don't mind I had to prefix our macros with `AZ_` so that they won't collide with anything defined by customer code.

File: sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/test/e2e/azure_core_with_curl_bodyBuffer.cpp]: I will close the created issue as not to be fixed since this is only tests which are not intented to validate the open file API.
FileBodyStream takes a file which was already opened and throws if read operation failed.

File: sdk/core/azure-core/src/datetime.cpp
Comment (RickWinter) [inline sdk/core/azure-core/src/datetime.cpp:37]: If you go this route everywhere we do the if/elif needs an /else #error.  Otherwise the code will compile but do nothing

Comment (antkmsft) [reply sdk/core/azure-core/src/datetime.cpp:37]: And these cases this is fine. If you don't use that functionality, you don't need a compilation error. The way these ifdefs are constructed, classes and function implementations will be missing, so you get the compilation and linker errors. But if you don't use that class, if you don't invoke that function, you are not blocked by `#error`s.

**Comment by antkmsft** [comment]
[11:41 AM] Rick Winter
    Added a comment about having an else in each of the cases.  Or you could do it in the platform.hpp.  Fail if we don't understand the platform
​[11:43 AM] Anton Kolesnyk
    Thank you Rick, I just replied
(1 liked)​[11:46 AM] Rick Winter
    As long as users get errors early on, I'm happy.  I commented yesterday as well that your change without this may have been ok as well since not haveing open(....) method would result in failures as well.  I prefer explicit checks for the platforms we support though
​[11:47 AM] Rick Winter
    I removed my block.  Thanks for addressing
​[11:49 AM] Anton Kolesnyk
    Thank you Rick! In cases where this may lead to UB, we have error. I also agree it is better to display errors early on if possible. But having them in all cases I think is extreme - you include stream header, you don't use file stream, but you get compilation error you have to address. If there is no error, as I have in my PR, you simply don't get the FileStream class, and if you don't attempt to use it, you won't notice that.
​[11:50 AM] Rick Winter
    Makes sense
​[11:51 AM] Anton Kolesnyk
    Something broke on github - I see you no longer requesting changes on the left, but at the bottom it still says "changes requested". I am going to try to dismiss that request then.

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:173]:
```suggestion
#ifdef defined(AZ_PLATFORM_POSIX)
```

File: sdk/core/azure-core/inc/azure/core/http/body_stream.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/body_stream.hpp:13]:
```suggestion
#ifdef defined(AZ_PLATFORM_POSIX)
```

File: sdk/core/azure-core/src/datetime.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:28]: In this case, we should add an `#else #error` because otherwise there will be a compiler error that there is no return statement, on other platforms. Or use the alternative approach where this function is only defined for windows/posix, based on our offline discussion. Either is fine.

**Pull request merged**
